### ISSUE TITLE
Database handling in graphback

### ISSUE THREAD

**Issue by ankitjena** [aerogear/graphback#56]
## Background
Currently `graphback-cli` is pretty much strict about postgres as the only database. Database creation in graphback is done using `knex` which supports a bunch of relational databases. 
 
`graphback-cli` has currently 3 main commands in it's workflow with other proposed commands yet to be finalized.
![steps](https://user-images.githubusercontent.com/23580602/59699296-ffc7ea80-920e-11e9-8fd4-3201f633b1b3.jpg)
- Init creates the project with model and config
- generate generates schema and resolvers
- db creates and manages database resources

## Discussion
Currently the moving parts which are dependent on database are the templates, resolver generation and database creation.

The user chooses from templates in the init command which are currently db specific(postgres).
- we can separate templates into different database specific ones, like `apollo-typescript-starter` to `apollo-typescript-pg`/`apollo-typescript-mongo`
- we can modify init command to inject db configs and docker-composes into templates with one skeleton.

The generate command generates `schema` and `resolvers`. The resolvers are currently implemented by a Knex resolver instance which generates resolvers which use `knex`. This is pluggable, so for it to support a no-SQL database like mongoDB would not be difficult.

#47 The `db` command will create db resources in local development server. Key point is to separate concerns regarding development and production.
Database to be created can be selected as we start to support different databases. But the concern of `db` command is to create local database.
- we can create a config file during init which store db config and accordingly create a docker-compose. But this will also have template as another moving piece which will be needed to configure.
- we should support addition of new tables, new columns(the current implemetaion of graphback needs to be a modified a bit) but dropping columns is to be debated.

Finally, we should create a `graphback migrate` command that will generate database definitions.

## Points to note
- This is a development workflow. Dealing with production using `graphback` is to be ideated.

**Comment by wtrocki** [comment]
Worth adding that template has generated folder that is dynamic:
Example:
```
|-----generated
       |-----schema.graphql
       |-----resolvers.ts
|-----model
       |-----<your-model-name>.graphql
|-----src
       |-----config
             |-----config.ts
       |-----db.ts
       |-----index.ts
       |-----mapping.ts
|-----package.json
|-----Dockerfile
|-----docker-compose.yml
|-----package-lock.json
|-----tslint.json
|-----README.md
```

Once the project is initialized users can possibly tweak all apart from the generated folder. We would really lose control over the rest of the stack and this can evolve towards production. Initial deployment to production can really base off development ddl model so migrate command is not essential to deliver complete workflow. Development, in this case, can be a wider term, but this is really about the fact that for localhost or any remote cluster desired we will automatically adjust database structure to work with the current schema/resolvers.

Couple questions based on the model:

- Where and how do you see the actual database running? Do we ignore that and assume that users will run docker-compose manually?

- How about the fact that we need to run this server after generation. How do you see people doing actual development? 

With the latest release this will look like this:

1. Init project
2. Run docker compose with db
3. Generate resolvers
4. Generate DB 
5. Run server manually

-- Now I want to edit something in graphql types.

6. Generate resolvers again
7. Drop database
8. Apply database schema
9. Start server again

While performing local diff may require to refactor it will be cool to have `watch` command that will look for modifications in types and perform steps from 3 to 9 automatically

**Comment by ankitjena** [comment]
`watch` command would be real nice feature. @wtrocki What do you think about templates for now as I have mentioned above?

**Comment by wtrocki** [comment]
Can you explain how you would see this working. Is all code editable or we will have runtime library that will setup db and it could be installed into vanilla templates.

**Comment by ankitjena** [comment]
My approach would be to separate db specific templates and have a attribute that helps us choose config and docker-compose automatically for the template. If a user chooses a postgres template then we will inject postgres config and accordingly for different databases. The user can further edit these files according to usecases.

**Comment by wtrocki** [comment]
Agreed with everything. There is no need for follow up issue as we have all done

**Comment by wtrocki** [comment]
I think it will be worth to investigate prisma2 for this task to see how we can avoid writing this complex subsystem and do database migrations.

**Comment by wtrocki** [comment]
Let's leave it open and provide more details once we finish spikes

**Comment by wtrocki** [comment]
Really good discussion here. We need to get back to the updated idea of how database changes can be handled. Without bringing the tooling that can be used I think we need to think if we can 
build state aware generation process.

Our project would have this state from previous generation, this will allow us to calculate diff and then see diff on the database to deliver sql statements. If we do that separately we can benefit from large set of the open source libraries that do database management right. 

So TL:DR idea is to detach database management problem from graphql problem by collecting 
schema changes and feeding them back to SQL Diff engine that can both apply changes directly or make migration scripts
